src/Microsoft.AspNetCore.Routing/Matchers/AsciiKeyedJumpTable.cs

```diff
@@ -0,0 +1,161 @@
+            {
+                if (entries[i].text.Length == 0)
+                {
+                    return false;
+                }
+
+                if (!IsAscii(entries[i].text))
+                {
+                    return false;
+                }
+                
+                var first = ToUpperAscii(entries[i].text[0]);
+                if (first < 'A' || first > 'Z')
+                {
+                    // Not a letter
+                    return false;
+                }
+
+                if (!map.TryGetValue(first, out var matches))
+                {
+                    matches = new List<(string text, int destination)>();
+                    map.Add(first, matches);
+                }
+
+                matches.Add(entries[i]);
+            }
+
+            var next = 0;
+            var ordered = new (string text, int destination)[entries.Count];
+            var indexes = new int[26 * 2];
+            for (var i = 0; i < 26; i++)
+            {
+                indexes[i * 2] = next;
+
+                var length = 0;
+                if (map.TryGetValue((char)('A' + i), out var matches))
+                {
+                    length += matches.Count;
+                    for (var j = 0; j < matches.Count; j++)
+                    {
+                        ordered[next++] = matches[j];
+                    }
+                }
+
+                indexes[i * 2 + 1] = length;
+            }
+
+            result = new AsciiKeyedJumpTable(@default, exit, ordered, indexes);
+            return true;
+        }
+
+        private readonly int _default;
+        private readonly int _exit;
+        private readonly (string text, int destination)[] _entries;
+        private readonly int[] _indexes;
+
+        private AsciiKeyedJumpTable(
+            int @default,
+            int exit,
+            (string text, int destination)[] entries,
+            int[] indexes)
+        {
+            _default = @default;
+            _exit = exit;
+            _entries = entries;
+            _indexes = indexes;
+        }
+
+        public unsafe override int GetDestination(string path, PathSegment segment)
+        {
+            if (segment.Length == 0)
+            {
+                return _exit;
+            }
+
+            var c = path[segment.Start];
+            if ((c & ~0x7F) != 0)
+            {
+                return _default;
+            }
+
+            c = (char)(c & 0x5F);
+            if (c < 'A' || c > 'Z')
+            {
+                // Character is non-ASCII or not a letter. Since we know that all of the entries are ASCII
+                // and begin with a letter this is not a match.
+                return _default;
+            }
+
+            var offset = (c - 'A') * 2;
+            var start = _indexes[offset];
+            var length = _indexes[offset + 1];
+
+            for (var i = start; i < start + length; i++)
+            {
+                ref var entry = ref _entries[i];
```

Comment:
Is this ref for performance? Worth commenting?

Reply:
meta: should we just put a blanket "we did the perf, you should not refactor this code unless you have a reason" on all of the perf sensitive code?

The explanation for most things will be that we benchmarked it and did the fastest thing.

Reply:
I've removed this since it's not actually the optimal thing.